### PR TITLE
reimplement errname lookup.

### LINKED ISSUE #31: re-implement errname resolution
#27 dropped execFile in favor of using spawn directly, so we've lost some features.

One is whatever these two lines do:

1. https://github.com/nodejs/node/blob/31600735f40e02935b936802ff42d66d49e9769b/lib/child_process.js#L8
2. https://github.com/nodejs/node/blob/31600735f40e02935b936802ff42d66d49e9769b/lib/child_process.js#L204

Can we reliably just do `process.binding('uv')`? Or is that asking for problems. I know it means importing some native lib, but not much beyond that.

Maybe we look back all the way to Node `0.12` and make sure it was still imported and used the same back then?

### PR DESCRIPTION
**Pull request by jamestalmage**
Fixes #31

I can not for the life of me figure out how to create a test that actually triggers the affected code in `index.js`. I can't create a negative error code in a way that causes `!err` to be true.

### CODE REVIEW

File: lib/errname.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+	return 'UNKNOWN CODE: ' + code;
+}
+
+module.exports = function getErrname(code) {
+	return errname(uv, code);
+};
+
+// used for testing the fallback behavior.
+module.exports._test = errname;
```
Comment (sindresorhus) [inline lib/errname.js]: Maybe we should have a convention of `__test__` for when you have to attach something for testing?

File: lib/errname.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+// If this ever stops working. See: https://github.com/sindresorhus/execa/issues/31#issuecomment-215939939 for another possible solution.
+var uv;
+
+try {
+	uv = process.binding('uv');
+	if (typeof uv.errname !== 'function') {
+		throw new Error('uv.errname is not a function');
+	}
+} catch (e) {
+	console.error('execa/lib/errname: unable to establish process.binding("uv")', e);
```
Comment (sindresorhus) [inline lib/errname.js]: `console.error('execa/lib/errname: unable to establish process.binding('uv')', e);`


and `e` => `err`

File: lib/errname.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+//   https://github.com/nodejs/node/pull/2768
+//
+// However, we need the 'uv' binding for errname support.
+// This is a defensive wrapper around it so `execa` will not fail entirely if it stops working someday.
+//
+// If this ever stops working. See: https://github.com/sindresorhus/execa/issues/31#issuecomment-215939939 for another possible solution.
+var uv;
+
+try {
+	uv = process.binding('uv');
```
Comment (sindresorhus) [inline lib/errname.js]: empty line here

File: test/errname.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+// simulates failure to capture process.binding('uv');
+function fallback(code) {
+	return errname._test(null, code);
+}
+
+function makeTests(name, m, expected) {
+	test(name, t => {
+		// throws >= 0
+		t.throws(() => m(0), /err >= 0/);
```
Comment (sindresorhus) [inline test/errname.js]: I don't see `m` imported anywhere.

Comment (jamestalmage) [inline test/errname.js]: It's a parameter to `makeTests`

**Comment by jamestalmage** [comment]
Updated.

**Comment by jamestalmage** [comment]
Hmm. Do you think there's more on `stderr` or `stdout` that's not being printed for [this error](https://ci.appveyor.com/project/sindresorhus/execa/build/55/job/ni8woa6980dftrhe#L808)?

File: test/errname.js [hunk 5]
```diff
@@ -1,9 +1,12 @@
+import os from 'os';
 import test from 'ava';
 import errname from '../lib/errname';
 
+const isWin = os.platform() === 'win32';
```
Comment (sindresorhus) [inline test/errname.js]:
```js
const isWin = process.platform === 'win32';
```

**Comment by sindresorhus** [comment]
Not really sure what you're asking.

File: test/errname.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import os from 'os';
+import test from 'ava';
+import errname from '../lib/errname';
+
+const isWin = os.platform() === 'win32';
+
+// simulates failure to capture process.binding('uv');
+function fallback(code) {
```
Comment (kevva) [inline test/errname.js]: `const fallback = code = {` looks nicer imo.

Comment (sindresorhus) [inline test/errname.js]: @kevva Join the discussion in https://github.com/sindresorhus/eslint-plugin-xo/issues/3 ;)

File: test/errname.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import os from 'os';
+import test from 'ava';
+import errname from '../lib/errname';
+
+const isWin = os.platform() === 'win32';
```
Comment (sindresorhus) [inline test/errname.js]:
```js
const isWin = process.platform === 'win32';
```

**Comment by jamestalmage** [comment]
> Not really sure what you are asking

I couldn't figure out why there was a test error. Turns out I was testing on Node 4 locally when I thought I had switched to Node 0.12.

It seems `uv.errname(-2)` causes a `process.exit` on Windows Node `0.12`.

I'm not even sure how to get negative error codes in Windows (or at all really).

**Comment by jamestalmage** [comment]
I think this is ready for merge. Not sure what to do about the `0.12` issue, but the JS code for the part of `execFile` that calls `uv.errname` is the same all the way back to `0.12`, so I don't think it's a problem.

**Comment by sindresorhus** [comment]
> Not sure what to do about the 0.12 issue

Nothing. We're doing exactly what Node.js is doing and I personally couldn't care less about 0.12.

**Comment by sindresorhus** [comment]
👍

**Pull request merged**